contracts/ERC721Collateralizer.sol

```diff
@@ -0,0 +1,426 @@
+                agreementId,
+                block.timestamp
+            ) > termsContract.getValueRepaidToDate(agreementId)
+        );
+
+        // Determine beneficiary of the collateral to be seized.
+        address beneficiary = debtRegistry.getBeneficiary(agreementId);
+
+        // Instantiate an instance of the token, in order to initiate a transfer.
+        ERC721 erc721token = ERC721(collateralTokenAddress);
+
+        // Apply a hack for CryptoKitties, granting transfer approval to this contract itself.
+        if (cryptoKittiesContract == collateralTokenAddress) {
+            erc721token.approve(address(this), collateralTokenID);
+        }
+
+        // Transfer the collateral this contract was holding in escrow to beneficiary.
+        erc721token.transferFrom(address(this), beneficiary, collateralTokenID);
+
+        // Ensure that transfer has succeeded.
+        require(erc721token.ownerOf(collateralTokenID) == beneficiary);
+
+        // Log the seizure event.
+        CollateralSeized(
+            agreementId,
+            beneficiary,
+            collateralTokenAddress,
+            collateralTokenID
+        );
+    }
+
+    /**
+     * Adds an address to the list of agents authorized
+     * to invoke the `collateralize` function.
+     */
+    function addAuthorizedCollateralizeAgent(address agent)
+        public
+        onlyOwner
+    {
+        collateralizationPermissions.authorize(agent, CONTEXT);
+    }
+
+    /**
+     * Removes an address from the list of agents authorized
+     * to invoke the `collateralize` function.
+     */
+    function revokeCollateralizeAuthorization(address agent)
+        public
+        onlyOwner
+    {
+        collateralizationPermissions.revokeAuthorization(agent, CONTEXT);
+    }
+
+    /**
+    * Returns the list of agents authorized to invoke the 'collateralize' function.
+    */
+    function getAuthorizedCollateralizeAgents()
+        public
+        view
+        returns (address[])
+    {
+        return collateralizationPermissions.getAuthorizedAgents();
+    }
+
+    /**
+     * Unpacks collateralization-specific parameters from their tightly-packed
+     * representation in a terms contract parameter string.
+     */
+    function unpackCollateralParametersFromBytes(bytes32 parameters)
+        public
+        pure
+        returns (bool, uint, uint)
+    {
+        // We use use a flag, 0 or 1, to specify whether the given collateral token's contract
+        // implements the Enumerable extension. If so, we can refer to a token by its index instead
+        // of its ID to conserve space. To retrieve the flag, we take only the value of the specific
+        // parameter that encodes it, and then shift that value 26 places to the right - thereby
+        // ignoring all other parameters.
+        bytes32 isEnumerableShifted =
+            parameters & 0x0000000000000000000000000000000000000100000000000000000000000000;
```

Comment:
If I'm not mistaken, we're still wasting 3 bits of space here that could used to give us, say, 8x as many possible `tokenRefs`.  This is because, even though we're AND'ing the half-byte with a binary `0001`, we're still not using those 3 left bits for anything.  Would make more sense IMO to put the binary flag on the far right of the string, so that the left three bits can contribute to the max-size of `tokenRef`